// popmotion/lib/simple-actions/Action.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _renderLoop = require('../framesync/render-loop');

var _calc = require('../inc/calc');

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Action = function () {
  function Action(props) {
    _classCallCheck(this, Action);

    this.props = _extends({}, props);
    this.current = 0;
    this.velocity = 0;
    this.update = this.update.bind(this);
  }

  Action.prototype.addChild = function addChild(action) {
    this.children.push(action);
  };

  Action.prototype.start = function start() {
    (0, _renderLoop.onFrameUpdate)(this.update);

    if (this.props.from === undefined && this.props.value) {
      this.props.from = this.props.value.getCurrent();
    }

    if (this.onStart) {
      fireOnStart(this);
    }

    return this;
  };

  Action.prototype.stop = function stop() {
    (0, _renderLoop.cancelOnFrameUpdate)(this.update);

    if (this.onStart) {
      fireOnStop(this);
    }

    return this;
  };

  Action.prototype.complete = function complete() {
    if (this.onComplete) {
      this.onComplete(this);
    }

    return this;
  };

  Action.prototype.update = function update(framestamp, elapsed) {
    var current = this.getCurrent();
    var velocity = this.getVelocity();

    if (this.onUpdate) {
      this.current = this.onUpdate(framestamp, elapsed);
      this.velocity = (0, _calc.speedPerSecond)(this.current - current, elapsed);

      if (this.velocity) {
        // tell renderer to render
      }
    }

    // Check if action is complete, if not, fire on the next frame
    if (this.isActionComplete && this.isActionComplete()) {
      this.complete();
    } else {
      (0, _renderLoop.onFrameUpdate)(this.update);
    }

    this.lastUpdated = framestamp;
  };

  Action.prototype.set = function set(props) {
    this.props = _extends({}, this.props, props);
    return this;
  };

  Action.prototype.get = function get(key) {
    return this.props[key];
  };

  Action.prototype.getCurrent = function getCurrent() {
    return this.current;
  };

  Action.prototype.getVelocity = function getVelocity() {
    return this.velocity;
  };

  return Action;
}();

;

exports.default = Action;
