// popmotion/lib/simple-actions/clock.js
'use strict';

exports.__esModule = true;
exports.default = clock;

var _Action2 = require('./Action');

var _Action3 = _interopRequireDefault(_Action2);

var _utils = require('../inc/utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var MAX_ELAPSED = 33;

var Clock = function (_Action) {
  _inherits(Clock, _Action);

  function Clock() {
    _classCallCheck(this, Clock);

    return _possibleConstructorReturn(this, _Action.apply(this, arguments));
  }

  Clock.prototype.onStart = function onStart() {
    this.current = (0, _utils.currentTime)();
  };

  Clock.prototype.onUpdate = function onUpdate(framestamp) {
    var elapsed = Math.min(framestamp - this.current, MAX_ELAPSED) * this.dilate;
    this.current += elapsed;
  };

  return Clock;
}(_Action3.default);

Clock.defaultProps = {
  dilate: 1
};

function clock(props) {
  return new Clock(props);
};

function flatten() {}

// // http://www.gamasutra.com/view/feature/131863/animation_blending_achieving_.php
// class CompositeAction extends Action {
//   constructor(props) {
//     super(props);
//     this.children = [];
//     this.numChildren = 0;
//     this.flattenList = [];
//   }

//   addAction(action) {
//     this.children.push({
//       action,
//       force,
//       weight: 0
//     });

//     this.numChildren++;
//   }

//   removeAction(action) {

//   }

//   onUpdate(framestamp) {
//     const numChildren = this.children.length;

//     for (let i = 0; i < numChildren; i++) {
//       this.children[i].onUpdate(framestamp);
//     }

//     flatten(this);

//     //blend
//     let currentWeight = 0;

//     for (i = 0; i < this.flattenList.length; i++) {
//       const item = this.flattenList[i];
//       if (item.blendWeight) {
//         currentWeight += item.blendWeight;
//         const interpolation = item.blendWeight / currentWeight;
//         blendChannels(buffer, item, interpolation);
//       }
//     }

//     function blendTwoTweens(a, b) {
//       const aEasedProgress
//       const bEasedProgress
//       const [aEasedProgress, 1] -> [0, 1]
//       const [bEasedProgress, 1] -> [0, 1]
//       const 
//     }
//   }
// }
