// popmotion/lib/simple-actions/group.js
'use strict';

exports.__esModule = true;
exports.default = actionGroup;

var _Action2 = require('./Action');

var _Action3 = _interopRequireDefault(_Action2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

group();

tween(opts);

function tween(props) {

  return function () {};
};

function actionGroup(actions) {
  return new ActionGroup({ actions: actions });
};

var TransferVelocity = function (_Action) {
  _inherits(TransferVelocity, _Action);

  function TransferVelocity() {
    _classCallCheck(this, TransferVelocity);

    return _possibleConstructorReturn(this, _Action.apply(this, arguments));
  }

  TransferVelocity.prototype.onStart = function onStart() {
    this.numActions = this.props.actions.length;
  };

  TransferVelocity.prototype.onUpdate = function onUpdate(framestamp, elapsed) {
    var actions = this.props.actions;

    var velocity = 0;

    for (var i = 0; i < this.numActions; i++) {
      var action = actions[i];

      if (i !== 0) {
        action.velocity = velocity;
      }

      action.onUpdate(framestamp, elapsed);
      velocity = action.getVelocity();
    }

    this.props.action();
  };

  return TransferVelocity;
}(_Action3.default);
