// popmotion/lib/simple-actions/physics.js
'use strict';

exports.__esModule = true;
exports.default = physics;

var _Action3 = require('./Action');

var _Action4 = _interopRequireDefault(_Action3);

var _calc = require('../inc/calc');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Physics = function (_Action) {
  _inherits(Physics, _Action);

  function Physics() {
    _classCallCheck(this, Physics);

    return _possibleConstructorReturn(this, _Action.apply(this, arguments));
  }

  Physics.prototype.onUpdate = function onUpdate(_ref) {
    var elapsed = _ref.elapsed;
    var velocity = this.props.velocity;

    // Apply acceleration

    velocity += (0, _calc.speedPerFrame)(this.acceleration, elapsed);

    // Apply friction
    velocity *= Math.pow(1 - this.friction, elapsed / 100);

    // Apply latest velocity
    this.current += (0, _calc.speedPerFrame)(velocity, elapsed);
  };

  return Physics;
}(_Action4.default);

var Spring = function (_Action2) {
  _inherits(Spring, _Action2);

  function Spring() {
    _classCallCheck(this, Spring);

    return _possibleConstructorReturn(this, _Action2.apply(this, arguments));
  }

  Spring.prototype.onUpdate = function onUpdate(_ref2) {
    var elapsed = _ref2.elapsed;

    var distanceToTarget = this.to - this.current;
    velocity += distanceToTarget * (0, _calc.speedPerFrame)(this.spring, elapsed);
  };

  return Spring;
}(_Action4.default);

Physics.valueProps = ['velocity', 'current'];

Physics.defaultProps = {
  acceleration: 0,
  spring: 0,
  to: 0
};

function physics(props) {
  return new Physics(props);
};
