// popmotion/lib/simple-actions/track.js
'use strict';

exports.__esModule = true;
exports.default = track;

var _Action3 = require('./Action');

var _Action4 = _interopRequireDefault(_Action3);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Track = function (_Action) {
  _inherits(Track, _Action);

  function Track() {
    _classCallCheck(this, Track);

    return _possibleConstructorReturn(this, _Action.apply(this, arguments));
  }

  Track.prototype.onStart = function onStart() {
    this.inputOrigin = this.input.getCurrent();
  };

  Track.prototype.onUpdate = function onUpdate(_ref) {
    var current = _ref.current;

    this.offset = this;
    this.current = this.from + this.inputOrigin;

    if (this.smooth) {
      this.current = smooth(this.current, current, elapsed, this.smooth);
    }
  };

  return Track;
}(_Action4.default);

Track.defaultProps = {
  from: 0,
  smooth: 0
};

function track(props) {
  return new Track(props);
};

var Anticipate = function (_Action2) {
  _inherits(Anticipate, _Action2);

  function Anticipate() {
    _classCallCheck(this, Anticipate);

    return _possibleConstructorReturn(this, _Action2.apply(this, arguments));
  }

  Anticipate.prototype.onUpdate = function onUpdate() {
    doThis(props);
  };

  return Anticipate;
}(_Action4.default);

new Anticipate({});
