// popmotion/lib/simple-actions/tween.js
'use strict';

exports.__esModule = true;
exports.default = tween;

var _Action2 = require('./Action');

var _Action3 = _interopRequireDefault(_Action2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Tween = function (_Action) {
  _inherits(Tween, _Action);

  function Tween() {
    _classCallCheck(this, Tween);

    return _possibleConstructorReturn(this, _Action.apply(this, arguments));
  }

  Tween.prototype.onStart = function onStart() {
    this.elapsed = 0;
  };

  Tween.prototype.onUpdate = function onUpdate(framestamp, elapsed) {
    var progress = restrict(getProgressFromValue(elapsed - this.delay, 0, this.duration), 0, 1);

    // Step progress if we're stepping
    if (this.steps) {
      progress = stepProgress(progress, this.steps);
    }

    // Ease progress
    this.current = ease(progress, this.from, this.stop, this.ease);
  };

  return Tween;
}(_Action3.default);

Tween.defaultProps = {
  ease: Easing.easeOut,
  elapsed: 0,
  steps: 0,
  from: 0,
  to: 0
};

function tween(props) {
  return new Tween({});
};
