// popmotion/lib/framesync/index.js
'use strict';

exports.__esModule = true;
exports.currentFrameTimestamp = exports.timeSinceLastFrame = exports.cancelOnFrameEnd = exports.cancelOnFrameRender = exports.cancelOnFrameUpdate = exports.cancelOnFrameStart = exports.onFrameEnd = exports.onFrameRender = exports.onFrameUpdate = exports.onFrameStart = undefined;

var _onNextFrame = require('./on-next-frame');

var _onNextFrame2 = _interopRequireDefault(_onNextFrame);

var _createRenderStep = require('./create-render-step');

var _createRenderStep2 = _interopRequireDefault(_createRenderStep);

var _utils = require('../inc/utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var willRenderNextFrame = false;

/**
 * Maximum permitted ms since last frame
 * @type {Number}
 */
var MAX_ELAPSED = 20;

/**
 * Current framestamp
 * @type {Number}
 */
var currentFramestamp = (0, _utils.currentTime)();

var elapsed = 0;

/**
 * Factor to multiply `elapsed` by - 
 * 0.5 would be slow motion, 2 would be fast
 * @type {Number}
 */
var dilation = 1;

function startRenderLoop() {
  if (!willRenderNextFrame) {
    willRenderNextFrame = true;
    (0, _onNextFrame2.default)(processFrame);
  }
}

var frameStart = (0, _createRenderStep2.default)(startRenderLoop);
var frameUpdate = (0, _createRenderStep2.default)(startRenderLoop);
var frameRender = (0, _createRenderStep2.default)(startRenderLoop);
var frameEnd = (0, _createRenderStep2.default)(startRenderLoop);

function processFrame(framestamp) {
  willRenderNextFrame = false;
  elapsed = Math.max(Math.min(framestamp - currentFramestamp, MAX_ELAPSED), 1) * dilation;
  currentFramestamp = framestamp;

  frameStart.process();
  frameUpdate.process();
  frameRender.process();
  frameEnd.process();
}

var onFrameStart = exports.onFrameStart = frameStart.schedule;
var onFrameUpdate = exports.onFrameUpdate = frameUpdate.schedule;
var onFrameRender = exports.onFrameRender = frameRender.schedule;
var onFrameEnd = exports.onFrameEnd = frameEnd.schedule;
var cancelOnFrameStart = exports.cancelOnFrameStart = frameStart.cancel;
var cancelOnFrameUpdate = exports.cancelOnFrameUpdate = frameUpdate.cancel;
var cancelOnFrameRender = exports.cancelOnFrameRender = frameRender.cancel;
var cancelOnFrameEnd = exports.cancelOnFrameEnd = frameEnd.cancel;

var timeSinceLastFrame = exports.timeSinceLastFrame = function () {
  return elapsed;
};
var currentFrameTimestamp = exports.currentFrameTimestamp = function () {
  return currentFramestamp;
};
