// popmotion/lib/inc/calc.js
'use strict';

exports.__esModule = true;
exports.stepProgress = exports.speedPerSecond = exports.speedPerFrame = exports.smooth = exports.restrict = exports.relativeValue = exports.random = exports.radiansToDegrees = exports.pointFromAngleAndDistance = exports.offset = exports.getProgressFromValue = exports.getValueFromProgress = exports.hypotenuse = exports.ease = exports.distance = exports.dilate = exports.degreesToRadians = exports.angle = undefined;

var _utils = require('./utils');

var ZERO_POINT = {
  x: 0,
  y: 0,
  z: 0
};

var distance1D = function (a, b) {
  return Math.abs(a - b);
};

/*
  Angle between points
  
  Translates the hypothetical line so that the 'from' coordinates
  are at 0,0
  
  @param [object]: X and Y coordinates of from point
  @param [object]: X and Y cordinates of to point
  @return [radian]: Angle between the two points in radians
*/
var angle = exports.angle = function (a) {
  var b = arguments.length <= 1 || arguments[1] === undefined ? ZERO_POINT : arguments[1];
  return radiansToDegrees(Math.atan2(a.x - b.x, a.y - b.y));
};

/*
  Convert degrees to radians
  
  @param [number]: Value in degrees
  @return [number]: Value in radians
*/
var degreesToRadians = exports.degreesToRadians = function (degrees) {
  return degrees * Math.PI / 180;
};

/*
  Dilate
  
  Change the progression between a and b according to dilation.
  
  So dilation = 0.5 would change
  
  a --------- b
  
  to
  
  a ---- b
  
  @param [number]: Previous value
  @param [number]: Current value
  @param [number]: Dilate progress by x
  @return [number]: Previous value plus the dilated difference
*/
var dilate = exports.dilate = function (a, b, dilation) {
  return a + (b - a) * dilation;
};

/*
  Distance
  
  Returns the distance between two n dimensional points.
  
  @param [object/number]: x and y or just x of point A
  @param [object/number]: (optional): x and y or just x of point B
  @return [number]: The distance between the two points
*/
var distance = exports.distance = function (a) {
  var b = arguments.length <= 1 || arguments[1] === undefined ? ZERO_POINT : arguments[1];

  // 1D dimensions
  if ((0, _utils.isNum)(a)) {
    return distance1D(a, b);

    // Multi-dimensional
  } else {
    var xDelta = distance1D(a.x, b.x);
    var yDelta = distance1D(a.y, b.y);
    var zDelta = (0, _utils.isNum)(a.z) ? distance1D(a.z, b.z) : 0;

    return Math.sqrt(Math.pow(xDelta, 2) + Math.pow(yDelta, 2) + Math.pow(zDelta, 2));
  }
};

/*
  Ease value within ranged parameters
  
  @param [number]: Progress between 0 and 1
  @param [number]: Value of 0 progress
  @param [number]: Value of 1 progress
  @param [string || function]: Name of preset easing
    to use or generated easing function
  @return [number]: Value of eased progress in range
*/
var ease = exports.ease = function (progress, from, to, ease) {
  var progressLimited = restrict(progress, 0, 1);
  var easedProgress = ease(progressLimited);

  return getValueFromProgress(easedProgress, from, to);
};

/*
  Hypotenuse
  
  Returns the hypotenuse, side C, given the lengths of sides A and B.
  
  @param [number]: Length of A
  @param [number]: Length of B
  @return [number]: Length of C
*/
var hypotenuse = exports.hypotenuse = function (a, b) {
  return Math.sqrt(a * a + b * b);
};

/*
  Value in range from progress
  
  Given a lower limit and an upper limit, we return the value within
  that range as expressed by progress (a number from 0-1)
  
  @param [number]: The progress between lower and upper limits expressed 0-1
  @param [number]: Lower limit of range
  @param [number]: Upper limit of range
  @return [number]: Value as calculated from progress within range (not limited within range)
*/
var getValueFromProgress = exports.getValueFromProgress = function (progress, from, to) {
  return -progress * from + progress * to + from;
};

/*
  Progress within given range
  
  Given a lower limit and an upper limit, we return the progress
  (expressed as a number 0-1) represented by the given value, and
  limit that progress to within 0-1.
  
  @param [number]: Value to find progress within given range
  @param [number]: Lower limit 
  @param [number]: Upper limit
  @return [number]: Progress of value within range as expressed 0-1
*/
var getProgressFromValue = exports.getProgressFromValue = function (value, from, to) {
  return (value - from) / (to - from);
};

/*
  Offset between two objects of numbers

  If property is found in b not present in a, it will return `0` for that prop.
  
  @param [Point]: First object
  @param [Point]: Second object
  @return [Offset]: Distance metrics between two points
*/
var offset = exports.offset = function (a, b) {
  var offset = {};

  for (var key in b) {
    if (b.hasOwnProperty(key)) {
      offset[key] = (0, _utils.hasProperty)(a, key) ? b[key] - a[key] : 0;
    }
  }

  return offset;
};

/*
  Point from angle and distance
  
  @param [object]: 2D point of origin
  @param [number]: Angle from origin
  @param [number]: Distance from origin
  @return [object]: Calculated 2D point
*/
var pointFromAngleAndDistance = exports.pointFromAngleAndDistance = function (origin, angle, distance) {
  angle = degreesToRadians(angle);

  return {
    x: distance * Math.cos(angle) + origin.x,
    y: distance * Math.sin(angle) + origin.y
  };
};

/*
  Convert radians to degrees
  
  @param [number]: Value in radians
  @return [number]: Value in degrees
*/
var radiansToDegrees = exports.radiansToDegrees = function (radians) {
  return radians * 180 / Math.PI;
};

/*
  Return random number between range
  
  @param [number] (optional): Output minimum
  @param [number] (optional): Output maximum
  @return [number]: Random number within range, or 0 and 1 if none provided
*/
var random = exports.random = function () {
  var min = arguments.length <= 0 || arguments[0] === undefined ? 0 : arguments[0];
  var max = arguments.length <= 1 || arguments[1] === undefined ? 1 : arguments[1];
  return Math.random() * (max - min) + min;
};

/*
  Calculate relative value
  
  Takes the operator and value from a string, ie "+=5", and applies
  to the current value to resolve a new target.
  
  @param [number]: Current value
  @param [string]: Relative value
  @return [number]: New value
*/
var relativeValue = exports.relativeValue = function (current, relative) {
  var newValue = current;
  var equation = relative.split('=');
  var operator = equation[0];

  var _findValueAndUnit = (0, _utils.findValueAndUnit)(equation[1]);

  var unit = _findValueAndUnit.unit;
  var value = _findValueAndUnit.value;


  value = parseFloat(value);

  switch (operator) {
    case '+':
      newValue += value;
      break;
    case '-':
      newValue -= value;
      break;
    case '*':
      newValue *= value;
      break;
    case '/':
      newValue /= value;
      break;
  }

  if (unit) {
    newValue += unit;
  }

  return newValue;
};

/*
  Restrict value to range
  
  Return value within the range of lowerLimit and upperLimit
  
  @param [number]: Value to keep within range
  @param [number]: Lower limit of range
  @param [number]: Upper limit of range
  @return [number]: Value as limited within given range
*/
var restrict = exports.restrict = function (value, min, max) {
  return Math.max(Math.min(value, max), min);
};

/*
  Framerate-independent smoothing

  @param [number]: New value
  @param [number]: Old value
  @param [number]: Frame duration
  @param [number] (optional): Smoothing (0 is none)
*/
var smooth = exports.smooth = function (newValue, oldValue, duration) {
  var smoothing = arguments.length <= 3 || arguments[3] === undefined ? 0 : arguments[3];
  return (0, _utils.toDecimal)(oldValue + duration * (newValue - oldValue) / Math.max(smoothing, duration));
};

/*
  Convert x per second to per frame velocity based on fps
  
  @param [number]: Unit per second
  @param [number]: Frame duration in ms
*/
var speedPerFrame = exports.speedPerFrame = function (xps, frameDuration) {
  return (0, _utils.isNum)(xps) ? xps / (1000 / frameDuration) : 0;
};

/*
  Convert velocity into velicity per second
  
  @param [number]: Unit per frame
  @param [number]: Frame duration in ms
*/
var speedPerSecond = exports.speedPerSecond = function (velocity, frameDuration) {
  return velocity * (1000 / frameDuration);
};

/*
  Create stepped version of 0-1 progress
  
  @param [number]: Current value
  @param [int]: Number of steps
  @return [number]: Stepped value
*/
var stepProgress = exports.stepProgress = function (progress, steps) {
  var segment = 1 / (steps - 1);
  var target = 1 - 1 / steps;
  var progressOfTarget = Math.min(progress / target, 1);

  return Math.floor(progressOfTarget / segment) * segment;
};
