// popmotion/lib/inc/filters.js
'use strict';

exports.__esModule = true;
exports.steps = exports.clamp = exports.clampOver = exports.clampUnder = exports.interpolate = exports.flow = exports.appendUnit = undefined;

var _calc = require('./calc');

/**
 * Append Unit
 * Creates a function that will append the unit to a given value
 * appendUnit('px')(20) -> '20px'
 * @param  {string} unit)
 * @return {number}
 */
var appendUnit = exports.appendUnit = function (unit) {
  return function (v) {
    return '' + v + unit;
  };
};

/**
 * Flow
 * Compose other transformers to run linearily
 * flow(min(20), max(40))
 * @param  {...functions} transformers
 * @return {function}
 */
var flow = exports.flow = function () {
  for (var _len = arguments.length, transformers = Array(_len), _key = 0; _key < _len; _key++) {
    transformers[_key] = arguments[_key];
  }

  var numTransformers = transformers.length;
  var i = 0;

  return function (v) {
    for (i = 0; i < numTransformers; i++) {
      v = transformers[i](v);
    }

    return v;
  };
};

var interpolate = exports.interpolate = function (input, output, rangeEasing) {
  var rangeLength = input.length;
  var finalIndex = rangeLength - 1;

  return function (v) {
    // If value outside minimum range, quickly return
    if (v <= input[0]) {
      return output[0];
    }

    // If value outside maximum range, quickly return
    if (v >= input[finalIndex]) {
      return output[finalIndex];
    }

    var i = 0;

    // Find index of range start
    for (; i < rangeLength; i++) {
      if (input[i] > v || i === finalIndex) {
        break;
      }
    }

    var progressInRange = (0, _calc.getProgressFromValue)(input[i], input[i + 1], v);
    var easedProgress = rangeEasing ? rangeEasing[i](progressInRange) : progressInRange;
    return (0, _calc.getValueFromProgress)(output[i], output[i + 1], easedProgress);
  };
};

/**
 * Clamp value between
 * Creates a function that will restrict a given value between `min` and `max`
 * @param  {number} min
 * @param  {number} max
 * @return {number}
 */
var clampUnder = exports.clampUnder = function (max) {
  return function (v) {
    return Math.min(v, max);
  };
};
var clampOver = exports.clampOver = function (min) {
  return function (v) {
    return Math.max(v, min);
  };
};
var clamp = exports.clamp = function (min, max) {
  return flow(clampOver(min), clampUnder(max));
};

var steps = exports.steps = function (steps, min, max) {
  return function (v) {
    var progress = (0, _calc.getProgressFromValue)(min, max, v);
    return (0, _calc.stepProgress)(steps, progress);
  };
};
