// popmotion/lib/inc/timeline.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.default = timeline;

var _Tween = require('../actions/Tween');

var _Tween2 = _interopRequireDefault(_Tween);

var _presetEasing = require('../actions/easing/preset-easing');

var _presetEasing2 = _interopRequireDefault(_presetEasing);

var _calc = require('./calc');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*
  @param [array]
    Sequential array of tweens, each item can be a tween or definition obj:

    [
      new Tween(),
      stagger(),
      timeline(),
      {
        tween: new Tween(),
        at: 100,
        offset: "+=100"
      }
    ]
*/
var analyze = function (defs) {
  var timeline = [];
  var numDefs = defs.length;
  var currentPlayhead = 0;

  var _loop = function (i) {
    var def = defs[i];
    var defIsObj = def.tween ? true : false;
    var tween = defIsObj ? def.tween : def;

    if (defIsObj) {
      if (def.offset !== undefined) {
        currentPlayhead = (0, _calc.relativeValue)(currentPlayhead, def.offset);
      } else if (def.at !== undefined) {
        currentPlayhead = def.at;
      }
    }

    var duration = 0;
    for (var key in tween.values) {
      if (tween.values.hasOwnProperty(key)) {
        var value = tween.values[key];
        duration = Math.max(duration, value.duration);
      }
    }

    timeline.push({
      from: currentPlayhead,
      duration: duration,
      fire: function (time) {
        return tween.seekTime(time);
      }
    });

    currentPlayhead += duration;
  };

  for (var i = 0; i < numDefs; i++) {
    _loop(i);
  }

  return { totalTime: currentPlayhead, timeline: timeline };
};

var setTweens = function (_ref) {
  var timeline = _ref.timeline;
  var timelineLength = _ref.timelineLength;
  var values = _ref.values;
  var duration = _ref.duration;

  for (var i = 0; i < timelineLength; i++) {
    var _tween = timeline[i];
    var tweenTime = values.p.current * duration - _tween.from;

    if (tweenTime >= -50 && tweenTime <= _tween.duration + 50) {
      _tween.fire(tweenTime);
    }
  }
};

function timeline(def) {
  var props = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];

  var _analyze = analyze(def);

  var totalTime = _analyze.totalTime;
  var timeline = _analyze.timeline;


  return new _Tween2.default(_extends({
    ease: _presetEasing2.default.linear
  }, props, {
    duration: totalTime,
    values: {
      p: { from: 0, to: 1 }
    },
    timeline: timeline,
    timelineLength: timeline.length,
    onRender: setTweens
  }));
}
