// popmotion/lib/_actions/Action.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _Task2 = require('../task/Task');

var _Task3 = _interopRequireDefault(_Task2);

var _calc = require('../inc/calc');

var _utils = require('../inc/utils');

var _detect = require('../value-types/detect');

var _detect2 = _interopRequireDefault(_detect);

var _numericalValues = require('../inc/numerical-values');

var _numericalValues2 = _interopRequireDefault(_numericalValues);

var _detectAdapter = require('../inc/detect-adapter');

var _detectAdapter2 = _interopRequireDefault(_detectAdapter);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var NUM_NUMERICAL_VALUES = _numericalValues2.default.length;

var defaultRenderer = function (_ref) {
  var state = _ref.state;
  var adapter = _ref.adapter;
  var adapterData = _ref.adapterData;
  var element = _ref.element;
  return adapter(element, state, adapterData);
};

var convertIfShouldBeNumber = function (value) {
  return !isNaN(value) ? parseFloat(value) : value;
};

var Action = function (_Task) {
  _inherits(Action, _Task);

  function Action() {
    var props = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];

    _classCallCheck(this, Action);

    props.state = {};
    props.valueKeys = [];
    props.parentKeys = [];
    return _possibleConstructorReturn(this, _Task.call(this, props));
  }

  /*
    # Set Action properties
    ## Set user-defined Action properties
     @param [object]
    @return [Action]
  */


  Action.prototype.set = function set() {
    var props = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];

    this.values = this.values || {};

    var values = props.values;

    var propsToSet = _objectWithoutProperties(props, ['values']);

    var inheritable = {};

    // Set non-consumed properties
    _Task.prototype.set.call(this, propsToSet);

    // Detect correct `adapter` if none exists and `element` is being set
    if (this.element) {
      if (!this.adapter) {
        // Ducktypish check for Adapter
        this.adapter = (0, _detectAdapter2.default)(this.element);

        if (this.adapter.getElementData) {
          this.adapterData = this.adapter.getElementData(this.element);
        }
      }

      if (!this.onRender) {
        this.onRender = defaultRenderer;
      }
    }

    // Prime an object to inherit from, with only `value` properties
    for (var key in this.defaultValue) {
      if (this.defaultValue.hasOwnProperty(key)) {
        if (propsToSet.hasOwnProperty(key)) {
          inheritable[key] = propsToSet[key];
        } else if (this[key] !== undefined) {
          inheritable[key] = this[key];
        }
      }
    }

    // Update existing values with inheritable properties
    for (var _key in this.values) {
      if (this.values.hasOwnProperty(_key)) {
        this.values[_key] = _extends({}, this.values[_key], inheritable);
      }
    }

    // Update
    if (values) {
      this.setValues(values, inheritable);

      // Precompute number of value key and parent keys to avoid per-frame measurement
      this.numValueKeys = this.valueKeys.length;
      this.numParentKeys = this.parentKeys.length;
    }

    return this;
  };

  Action.prototype.setValues = function setValues(values, inherit) {
    // Iterate over all incoming values and set
    for (var key in values) {
      if (values.hasOwnProperty(key)) {
        var hasChildren = false;
        var children = {};

        // Merge into existing value or create new
        var valueAlreadyExists = this.values[key] !== undefined;
        var newValue = valueAlreadyExists ? _extends({}, this.values[key]) : _extends({}, inherit);

        // If values is not an object, assign value to default prop
        if (!(0, _utils.isObj)(values[key])) {
          newValue[this.defaultValueProp] = values[key];
        } else {
          newValue = _extends({}, newValue, values[key]);
        }

        // If we've got an adapter, get the current value
        if (newValue.current === undefined && this.adapter) {
          newValue.current = convertIfShouldBeNumber(this.adapter.get(this.element, key));
        }

        if (newValue.from === undefined && this.adapter) {
          newValue.from = newValue.current;
        }

        // Apply default value properties
        if (!valueAlreadyExists) {
          newValue = _extends({}, this.defaultValue, newValue);
        }

        // If we don't have a value type and we do have an Adapter, check for type with value key
        if (!newValue.type && this.adapter && this.adapter.checkValueType) {
          newValue.type = this.adapter.checkValueType(key);
        }

        // If we still don't have a value type and this is the first time we've set this value, check numerical values for strings and test
        if (!newValue.type && !this.values[key]) {
          newValue.type = (0, _detect2.default)(newValue);
        }

        // If we have a value type, handle. This is my least favourite part of Popmotion, so... enjoy.
        if (newValue.type) {
          for (var i = 0; i < NUM_NUMERICAL_VALUES; i++) {
            var propName = _numericalValues2.default[i];
            var valueProp = newValue[propName];

            // If this prop is a string and we have a splitter, split
            if (newValue.type.hasOwnProperty('split')) {
              var splitProp = (0, _utils.isString)(valueProp) ? newValue.type.split(valueProp) : {};

              for (var splitKey in splitProp) {
                if (splitProp.hasOwnProperty(splitKey)) {
                  var combinedKey = key + splitKey;

                  // If we don't have a child value for this key, make one
                  if (!children[combinedKey]) {
                    var defaultValue = newValue.type.defaultProps && newValue.type.defaultProps[splitKey] ? newValue.type.defaultProps[splitKey] : newValue.type.defaultProps || {};

                    children[combinedKey] = _extends({}, newValue, defaultValue, {
                      parent: key,
                      childKey: splitKey
                    });

                    delete children[combinedKey].type;
                  }

                  hasChildren = true;
                  children[combinedKey][propName] = parseFloat(splitProp[splitKey]);
                }
              }

              // If we have a template function, generate
              if (!newValue.template && newValue.type.template && (0, _utils.isString)(valueProp)) {
                newValue.template = newValue.type.template(valueProp);
              }
            } else if (newValue.type.defaultProps) {
              newValue = _extends({}, newValue, newValue.type.defaultProps);
            }

            if (valueProp !== undefined && newValue.type.parse) {
              newValue[propName] = newValue.type.parse(valueProp, newValue);
            }
          }
        } // End value type nonsense

        // Set `prev` to `current` for first frame after set
        newValue.prev = newValue.current;

        // If this value doesn't have children, add to valueKeys
        if (!hasChildren) {
          if (this.valueKeys.indexOf(key) === -1) {
            this.valueKeys.push(key);
          }

          // Or add to parentKeys
        } else {
          newValue.children = newValue.children || {};

          if (this.parentKeys.indexOf(key) === -1) {
            this.parentKeys.push(key);
          }

          this.setValues(children);
        }

        this.values[key] = newValue;
      }
    }
  };

  /*
    Decide whether this Action will render on next frame
     @param [Action]
    @param [number]
    @param [number]
    @return [boolean]: Return true to render
  */


  Action.prototype.willRender = function willRender(action, frameStamp, elapsed) {
    var hasChanged = false;

    // Check if base values have updated 
    for (var i = 0; i < this.numValueKeys; i++) {
      var key = this.valueKeys[i];
      var value = this.values[key];

      // Run transform function (if present)
      if (value.transform) {
        value.current = value.transform(value.current, key, this);
      }

      // Cap minimum
      if ((0, _utils.isNum)(value.min)) {
        value.current = Math.max(value.current, value.min);
      }

      // Cap maximum
      if ((0, _utils.isNum)(value.max)) {
        value.current = Math.min(value.current, value.max);
      }

      // Round number
      if (value.round) {
        value.current = Math.round(value.current);
      }

      value.frameChange = value.current - value.prev;

      // Update velocity
      if (!this.calculatesVelocity) {
        value.velocity = (0, _calc.speedPerSecond)(value.frameChange, elapsed);
      }

      // If this value has changed
      if (value.prev !== value.current) {
        hasChanged = true;
        value.prev = value.current;
      }

      // Append unit
      var valueForState = value.type && value.type.serialize ? value.type.serialize(value.current, value) : value.current;

      // Add to state if this is not a child vaue
      if (!value.parent) {
        this.state[key] = valueForState;
      } else {
        this.values[value.parent].children[value.childKey] = valueForState;
      }
    }

    // Update parent values
    for (var _i = 0; _i < this.numParentKeys; _i++) {
      var _key2 = this.parentKeys[_i];
      var _value = this.values[_key2];

      _value.current = _value.type.combine(_value.children, _value.template);

      this.state[_key2] = _value.current;
    }

    if (this.onFrame) {
      this.onFrame(this.state, this);
    }

    return this.onCleanup ? true : hasChanged;
  };

  Action.prototype.inherit = function inherit() {
    var props = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];
    var values = props.values;

    var propsToSet = _objectWithoutProperties(props, ['values']);

    var newAction = _Task.prototype.inherit.call(this, propsToSet);

    if (values) {
      newAction.set({ values: values });
    }

    return newAction;
  };

  Action.prototype.pause = function pause() {
    _Task.prototype.stop.call(this);
    return this;
  };

  Action.prototype.resume = function resume() {
    _Task.prototype.start.call(this);
    return this;
  };

  Action.prototype.toggle = function toggle() {
    return this.isActive ? this.pause() : this.resume();
  };

  Action.prototype.start = function start() {
    var values = this.values;
    _Task.prototype.start.call(this);

    for (var key in values) {
      if (values.hasOwnProperty(key)) {
        values[key].prev = values[key].origin = values[key].current;
      }
    }

    return this;
  };

  Action.extendDefaultValue = function extendDefaultValue(props) {
    return _extends({}, this.prototype.defaultValue, props);
  };

  Action.extendDefaultProps = function extendDefaultProps(props) {
    return _extends({}, this.prototype.defaultProps, props);
  };

  return Action;
}(_Task3.default);

Action.prototype.defaultValueProp = 'current';
Action.prototype.defaultValue = {
  current: 0,
  velocity: 0,
  round: false,
  min: undefined,
  max: undefined,
  transform: undefined
};

exports.default = Action;
