// popmotion/lib/_actions/Physics.js
'use strict';

exports.__esModule = true;

var _Action2 = require('./Action');

var _Action3 = _interopRequireDefault(_Action2);

var _calc = require('../inc/calc');

var _utils = require('../inc/utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Physics = function (_Action) {
  _inherits(Physics, _Action);

  function Physics() {
    _classCallCheck(this, Physics);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    var _this = _possibleConstructorReturn(this, _Action.call.apply(_Action, [this].concat(args)));

    _this.inactiveFrames = 0;
    _this.calculatesVelocity = true;
    return _this;
  }

  Physics.prototype.onUpdate = function onUpdate(physics, frameStamp, elapsed) {
    this.hasChanged = false;

    for (var i = 0; i < this.numValueKeys; i++) {
      var key = this.valueKeys[i];
      var value = this.values[key];
      var previousValue = value.prev;

      // Apply acceleration
      value.velocity += (0, _calc.speedPerFrame)(value.acceleration, elapsed);

      // Apply friction
      value.velocity *= Math.pow(1 - value.friction, elapsed / 100);

      // Apply spring
      if (value.spring && (0, _utils.isNum)(value.to)) {
        var distanceToTarget = value.to - value.current;
        value.velocity += distanceToTarget * (0, _calc.speedPerFrame)(value.spring, elapsed);
      }

      // Apply latest velocity
      value.current += (0, _calc.speedPerFrame)(value.velocity, elapsed);

      // Detect bounce
      if (value.min !== undefined && value.current < value.min || value.max !== undefined && value.current > value.max) {
        value.velocity *= -value.bounce;
      }

      // Check if value has changed
      if (value.current !== previousValue || Math.abs(value.velocity) >= value.stopSpeed || value.spring && value.current !== value.to) {
        this.hasChanged = true;
      }
    }
  };

  Physics.prototype.onFrameEnd = function onFrameEnd() {
    if (this.maxInactiveFrames !== Infinity) {
      this.inactiveFrames = this.hasChanged ? 1 : this.inactiveFrames + 1;

      if (this.inactiveFrames >= this.maxInactiveFrames) {
        this.complete();
      }
    }
  };

  return Physics;
}(_Action3.default);

Physics.prototype.defaultValueProp = 'velocity';
Physics.prototype.defaultValue = _Action3.default.extendDefaultValue({
  acceleration: 0, // [number]: Acceleration to apply to value, in units per second
  bounce: 0, // [number]: Factor to multiply velocity by on bounce
  spring: 0, // [number]: Spring strength during 'string'
  stopSpeed: 0.001, // [number]: Stop simulation under this speed
  friction: 0 // [number]: Friction to apply per frame, 0-1
});
Physics.prototype.defaultProps = _Action3.default.extendDefaultProps({
  maxInactiveFrames: 3
});

exports.default = Physics;
