// popmotion/lib/task/Task.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _loop = require('./loop');

var loop = _interopRequireWildcard(_loop);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } } /*
                                                                                                                                                            Base Task class for creating a task on the main render loop.
                                                                                                                                                          */


function cleanup() {
  this.onCleanup = undefined;
  loop.deactivate(this.id);
}

function activate() {
  this.onCleanup = cleanup;
}

var Task = function () {
  function Task(props) {
    _classCallCheck(this, Task);

    this.id = loop.getTaskId();
    this.isActive = false;

    if (this.defaultProps) {
      for (var key in this.defaultProps) {
        if (this.defaultProps.hasOwnProperty(key)) {
          this[key] = this.defaultProps[key];
        }
      }
    }

    this.set(props);
  }

  Task.prototype.set = function set(props) {
    for (var key in props) {
      if (props.hasOwnProperty(key)) {
        this[key] = props[key];
      }
    }

    return this;
  };

  Task.prototype.start = function start() {
    loop.activate(this.id, this);

    this.onActivateLoop = this.onCleanup = undefined;
    this.isComplete = false;

    if (this.onStart) {
      this.onStart(this);
    }

    return this;
  };

  Task.prototype.stop = function stop() {
    loop.deactivate(this.id);

    if (this.onStop) {
      this.onStop(this);
    }

    return this;
  };

  Task.prototype.once = function once() {
    loop.activate(this.id, this);
    this.onCleanup = undefined;
    this.onActivateLoop = activate;

    return this;
  };

  Task.prototype.complete = function complete() {
    this.stop();

    if (this.onComplete) {
      this.onComplete(this);
    }
  };

  /*
    # Extend this Process with new properties
    ## Returns new instance of this Process's `prototype` with existing and new properties
     @param [object] (optional)
    @return [Process]
  */


  Task.prototype.inherit = function inherit(props) {
    var id = this.id;

    var inheritedProps = _objectWithoutProperties(this, ['id']);

    return new this.constructor(_extends({}, inheritedProps, props));
  };

  return Task;
}();

exports.default = Task;
