// popmotion/lib/actions/Flow.js
'use strict';

exports.__esModule = true;

var _Action2 = require('../actions/Action');

var _Action3 = _interopRequireDefault(_Action2);

var _generateBlendCurve = require('./flow/generate-blend-curve');

var _generateBlendCurve2 = _interopRequireDefault(_generateBlendCurve);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/*
  Methods and properties to add to bound Actions
*/
var boundOnStart = function (action) {
  return action.flow.activateAction(action.id, action);
};
var boundOnStop = function (action) {
  return action.flow.deactivateAction(action.id);
};
var boundProps = function (flow) {
  return {
    flow: flow,
    isPriority: true,
    onActivate: boundOnStart,
    onDeactivate: boundOnStop
  };
};

var Flow = function (_Action) {
  _inherits(Flow, _Action);

  function Flow(props) {
    _classCallCheck(this, Flow);

    var _this = _possibleConstructorReturn(this, _Action.call(this, props));

    _this.activeActions = {};
    _this.numActiveActions = 0;
    return _this;
  }

  Flow.prototype.set = function set(props) {
    _Action.prototype.set.call(this, props);

    this.once();

    return this;
  };

  /*
    Bind Action to Actor
  */


  Flow.prototype.connect = function connect(action) {
    var inheritedAction = action.inherit();
    var newValues = {};
    var hasNewValues = false;

    // Create values on actor that don't exist
    for (var key in inheritedAction.values) {
      if (inheritedAction.values.hasOwnProperty(key) && !this.values.hasOwnProperty(key)) {
        newValues[key] = {};
        hasNewValues = true;
      }
    }

    if (hasNewValues) {
      this.set({ values: newValues });
    }

    inheritedAction.parentId = action.id;

    return inheritedAction.set(boundProps(this, inheritedAction));
  };

  /*
    Start Actor
     If Action is provided, bind it to this Actor and start
     @param (optional) [Action]
  */


  Flow.prototype.start = function start() {
    _Action.prototype.start.call(this);

    for (var key in this.activeActions) {
      if (this.activeActions.hasOwnProperty(key)) {
        var action = this.activeActions[key];
        if (!action.isActive) {
          action.start();
        }
      }
    }

    return this;
  };

  Flow.prototype.stop = function stop() {
    _Action.prototype.stop.call(this);

    for (var key in this.activeActions) {
      if (this.activeActions.hasOwnProperty(key)) {
        this.activeActions[key].stop();
      }
    }

    return this;
  };

  Flow.prototype.willRender = function willRender(actor, frameStamp, elapsed) {
    for (var i = 0; i < this.numValueKeys; i++) {
      var key = this.valueKeys[i];
      var value = this.values[key];
      var driver = value.numDrivers ? this.activeActions[value.drivers[0]] : false;
      var newCurrent = value.numDrivers ? driver.values[key].current : value.current;

      /**
       * TODO: replace with blend tree resolver
       * Additive motion
       * Bezier tween blend
       */
      if (value.blendCurve) {
        newCurrent = value.blendCurve();
      }

      value.current = newCurrent;
    }

    return _Action.prototype.willRender.call(this, actor, frameStamp, elapsed);
  };

  /*
    Add active actions
     @param [number]
    @param [Action]
  */


  Flow.prototype.activateAction = function activateAction(id, action) {
    this.activeActions[id] = action;
    this.numActiveActions++;

    for (var i = 0; i < action.numValueKeys; i++) {
      var key = action.valueKeys[i];
      var actionValue = action.values[key];
      var value = this.values[key];

      // If we're blending this action, and there's on already in progress
      if (action.blend && value.numDrivers && !value.blendCurve && value.drivers[0].prototype === action.prototype) {
        value.blendCurve = (0, _generateBlendCurve2.default)(this.activeActions[value.drivers[0]], action, value, key);
      } else if (!action.isScrubbing) {
        value.blendCurve = undefined;

        // Pass Actor value properties to Action
        if (actionValue.velocity === 0) {
          actionValue.velocity = value.velocity;
        }

        if (actionValue.from === undefined) {
          actionValue.from = actionValue.current = value.current;
        }
      }

      value.drivers = [id];
      value.numDrivers = value.drivers.length;
    }

    if (this.numActiveActions) {
      _Action.prototype.start.call(this);
    }
  };

  /*
    Remove active actions
     @param [number]
  */


  Flow.prototype.deactivateAction = function deactivateAction(id) {
    var action = this.activeActions[id];

    if (action) {
      for (var i = 0; i < action.numValueKeys; i++) {
        var key = action.valueKeys[i];
        var value = this.values[key];
        var driverIndex = value.drivers.indexOf(id);

        if (driverIndex !== -1) {
          value.drivers.splice(driverIndex, 1);
          value.numDrivers--;
        }
      }

      delete this.activeActions[id];
      this.numActiveActions--;
    }

    if (!this.numActiveActions && this.isActive) {
      _Action.prototype.stop.call(this);
    }
  };

  return Flow;
}(_Action3.default);

Flow.prototype.defaultValue = _Action3.default.extendDefaultValue({
  drivers: [],
  numDrivers: 0
});

exports.default = Flow;
