// popmotion/lib/actions/index.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _framesync = require('../framesync');

var _calc = require('../inc/calc');

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Action = function () {
  function Action(props) {
    _classCallCheck(this, Action);

    this.update = this.update.bind(this);

    this.props = _extends({}, this.constructor.defaultProps);

    this.setProps(props);

    this.current = props.current || 0;
  }

  Action.prototype.start = function start() {
    var _props = this.props;
    var onStart = _props.onStart;
    var _onStart = _props._onStart;
    var passive = _props.passive;


    if (!passive) {
      this.isActive = true;
      (0, _framesync.onFrameUpdate)(this.update);
    }

    if (this.onStart) this.onStart();
    if (onStart) onStart(this);
    if (_onStart) _onStart(this);

    return this;
  };

  Action.prototype.stop = function stop() {
    var _props2 = this.props;
    var onStop = _props2.onStop;
    var _onStop = _props2._onStop;
    var passive = _props2.passive;


    if (!passive) {
      this.isActive = false;
      (0, _framesync.cancelOnFrameUpdate)(this.update);
    }

    if (this.onStop) this.onStop();
    if (onStop) onStop(this);
    if (_onStop) _onStop(this);

    return this;
  };

  Action.prototype.complete = function complete() {
    var _props3 = this.props;
    var onComplete = _props3.onComplete;
    var _onComplete = _props3._onComplete;


    if (this.onComplete) this.onComplete();
    if (onComplete) onComplete(this);
    if (_onComplete) _onComplete(this);

    this.stop();

    return this;
  };

  Action.prototype.update = function update() {
    this.lastUpdated = (0, _framesync.timeSinceLastFrame)();
    this.prev = this.current;

    var _props4 = this.props;
    var onUpdate = _props4.onUpdate;
    var _onUpdate = _props4._onUpdate;
    var passive = _props4.passive;


    if (this.onUpdate) {
      this.current = this.onUpdate();
    }

    var filtered = this.get();
    if (onUpdate) onUpdate(filtered, this);
    if (_onUpdate) _onUpdate(filtered, this);

    if (!passive && this.isActive) {
      (0, _framesync.onFrameUpdate)(this.update);
    }

    if (this.isActionComplete && this.isActionComplete()) {
      this.complete();
    }

    return this;
  };

  Action.prototype.setProps = function setProps(props) {
    this.props = _extends({}, this.props, props);
  };

  Action.prototype.get = function get() {
    var filter = this.props.filter;

    return filter ? filter(this.current) : this.current;
  };

  Action.prototype.getVelocity = function getVelocity() {
    return (0, _calc.speedPerSecond)(this.prev - this.current, this.lastUpdated);
  };

  return Action;
}();

exports.default = Action;
