// popmotion/lib/actions/parallel.js
'use strict';

exports.__esModule = true;

var _ = require('./');

var _2 = _interopRequireDefault(_);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Parallel = function (_Action) {
  _inherits(Parallel, _Action);

  function Parallel() {
    _classCallCheck(this, Parallel);

    return _possibleConstructorReturn(this, _Action.apply(this, arguments));
  }

  Parallel.prototype.onStart = function onStart() {
    var _this2 = this;

    var actions = this.props.actions;

    this.numActiveActions = actions.length;

    actions.forEach(function (action) {
      action.setProps({
        _onStop: function () {
          return _this2.numActiveActions--;
        }
      }).start();
    });
  };

  Parallel.prototype.onStop = function onStop() {
    this.props.actions.forEach(function (action) {
      return action.stop();
    });
  };

  Parallel.prototype.addAction = function addAction(action) {
    var actions = this.props.actions;


    if (actions.indexOf(action) === -1) {
      actions.push(action);
    }
  };

  Parallel.prototype.isActionComplete = function isActionComplete() {
    return this.numActiveActions === 0;
  };

  return Parallel;
}(_2.default);

exports.default = function (actions) {
  return new Parallel({ actions: actions });
};
