// popmotion/lib/actions/Track.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _Action2 = require('./Action');

var _Action3 = _interopRequireDefault(_Action2);

var _Pointer = require('../input/Pointer');

var _Pointer2 = _interopRequireDefault(_Pointer);

var _calc = require('../inc/calc');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/*
  Scrape x/y coordinates from provided event

  @param [event]
  @return [object]
*/
var mouseEventToPoint = function (e) {
  return {
    x: e.pageX,
    y: e.pageY
  };
};

var touchEventToPoint = function (_ref) {
  var changedTouches = _ref.changedTouches;
  return {
    x: changedTouches[0].clientX,
    y: changedTouches[0].clientY
  };
};

var createPointer = function (e, preventDefault) {
  return e.touches ? new _Pointer2.default(touchEventToPoint(e), 'touchmove', touchEventToPoint, preventDefault) : new _Pointer2.default(mouseEventToPoint(e), 'mousemove', mouseEventToPoint, preventDefault);
};

var getActualEvent = function (e) {
  return e.originalEvent || e;
};

var Track = function (_Action) {
  _inherits(Track, _Action);

  function Track() {
    _classCallCheck(this, Track);

    return _possibleConstructorReturn(this, _Action.apply(this, arguments));
  }

  Track.prototype.start = function start(input) {
    var preventDefault = arguments.length <= 1 || arguments[1] === undefined ? true : arguments[1];

    _Action.prototype.start.call(this);

    if (input) {
      this.input = input.state ? input : createPointer(getActualEvent(input), preventDefault);
    }

    this.inputOffset = {};
    this.inputOrigin = _extends({}, this.input.state);
    this.input.start();

    return this;
  };

  Track.prototype.stop = function stop() {
    _Action.prototype.stop.call(this);
    this.input.stop();

    return this;
  };

  Track.prototype.onUpdate = function onUpdate(track, frameStamp, elapsed) {
    this.inputOffset = (0, _calc.offset)(this.inputOrigin, this.input.state);

    for (var i = 0; i < this.numValueKeys; i++) {
      var key = this.valueKeys[i];

      if (this.inputOffset.hasOwnProperty(key)) {
        var value = this.values[key];
        var inputProp = value.hasOwnProperty('watch') ? value.watch : key;

        if (value.direct) {
          value.current = this.input.state[inputProp];
        } else {
          value.current = value.from + this.inputOffset[inputProp];
        }

        // Smooth value if we have smoothing
        if (value.smooth) {
          value.current = (0, _calc.smooth)(value.current, value.prev, elapsed, value.smooth);
        }
      }
    }
  };

  return Track;
}(_Action3.default);

Track.prototype.defaultValueProp = 'watch';
Track.prototype.defaultValue = _Action3.default.extendDefaultValue({
  direct: false,
  from: 0
});

exports.default = Track;
