// popmotion/lib/actions/Tween.js
'use strict';

exports.__esModule = true;

var _Action2 = require('./Action');

var _Action3 = _interopRequireDefault(_Action2);

var _presetEasing = require('./easing/preset-easing');

var _presetEasing2 = _interopRequireDefault(_presetEasing);

var _utils = require('../inc/utils');

var _calc = require('../inc/calc');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var COUNT = 'Count';
var NEXT_STEPS = {
  loop: 'restart',
  yoyo: 'reverse',
  flip: 'flipValues'
};

var Tween = function (_Action) {
  _inherits(Tween, _Action);

  function Tween() {
    _classCallCheck(this, Tween);

    return _possibleConstructorReturn(this, _Action.apply(this, arguments));
  }

  Tween.prototype.start = function start() {
    this.elapsed = 0;
    this.flipCount = this.yoyoCount = this.loopCount = 0;
    this.isScrubbing = false;

    // Set default `from` if none set
    if (!this.flow) {
      for (var i = 0; i < this.numValueKeys; i++) {
        var key = this.valueKeys[i];
        var value = this.values[key];
        if (value.from === undefined) {
          value.from = 0;
        }
      }
    }

    return _Action.prototype.start.call(this);
  };

  Tween.prototype.onUpdate = function onUpdate(tween, frameStamp, elapsed) {
    var progressTarget = this.playDirection === 1 ? 1 : 0;

    this.ended = true;

    if (!this.isScrubbing) {
      this.elapsed += elapsed * this.dilate * this.playDirection;
    }

    for (var i = 0; i < this.numValueKeys; i++) {
      var key = this.valueKeys[i];
      var value = this.values[key];

      var progress = (0, _calc.restrict)((0, _calc.getProgressFromValue)(this.elapsed - value.delay, 0, value.duration), 0, 1);

      // Mark Tween as NOT ended if still in progress
      if (progress !== progressTarget) {
        this.ended = false;
      }

      // Step progress if we're stepping
      if (value.steps) {
        progress = (0, _calc.stepProgress)(progress, value.steps);
      }

      // Ease progress
      value.current = (0, _calc.ease)(progress, value.from, value.to, value.ease);
    }
  };

  Tween.prototype.onFrameEnd = function onFrameEnd() {
    if (this.ended) {
      var stepTaken = false;

      for (var key in NEXT_STEPS) {
        if (NEXT_STEPS.hasOwnProperty(key)) {
          if ((0, _utils.isNum)(this[key]) && this[key] > this[key + COUNT]) {
            this[key + COUNT]++;
            stepTaken = true;
            this[NEXT_STEPS[key]]();
          }
        }
      }

      if (!stepTaken) {
        this.complete();
      }
    }
  };

  Tween.prototype.flipValues = function flipValues() {
    var values = this.values;

    this.elapsed = this.duration - this.elapsed;

    for (var key in values) {
      if (values.hasOwnProperty(key)) {
        var value = values[key];
        var _ref = [value.from, value.to];
        value.to = _ref[0];
        value.from = _ref[1];
      }
    }

    return this;
  };

  Tween.prototype.reverse = function reverse() {
    this.playDirection *= -1;
    return this;
  };

  Tween.prototype.restart = function restart() {
    this.elapsed = this.playDirection === 1 ? 0 : this.duration;
    this.started = (0, _utils.currentTime)();
    return this;
  };

  Tween.prototype.seek = function seek(progress) {
    this.seekTime(this.duration * progress);
    return this;
  };

  Tween.prototype.seekTime = function seekTime(elapsed) {
    if (!this.isActive || this.isScrubbing) {
      this.once();
      this.isScrubbing = true;
    }

    this.elapsed = elapsed;

    return this;
  };

  return Tween;
}(_Action3.default);

Tween.prototype.defaultValueProp = 'to';
Tween.prototype.defaultValue = _Action3.default.extendDefaultValue({
  delay: 0,
  duration: 300,
  ease: _presetEasing2.default.easeOut,
  elapsed: 0,
  steps: 0,
  to: 0,
  round: false
});
Tween.prototype.defaultProps = _Action3.default.extendDefaultProps({
  blend: false,
  dilate: 1,
  loop: 0,
  yoyo: 0,
  flip: 0,
  loopCount: 0,
  yoyoCount: 0,
  flipCount: 0,
  playDirection: 1,
  isScrubbing: false,
  ended: false,
  elapsed: 0
});

exports.default = Tween;
