// popmotion/lib/render/render.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _renderLoop = require('../framesync/render-loop');

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Render = function () {
  function Render(onChange, valueMap) {
    _classCallCheck(this, Render);

    this.values = _extends({}, valueMap);
    this.state = {};
    this.onChange = onChange;

    for (var key in valueMap) {
      this.register(valueMap[key]);
    }
  }

  Render.prototype.register = function register(value) {
    value.subscribe(this.invalidate);
  };

  Render.prototype.invalidate = function invalidate() {
    if (!this.hasChanged) {
      (0, _renderLoop.onFrameRender)(this.render);
    }

    this.hasChanged = true;
  };

  Render.prototype.render = function render() {
    this.hasChanged = false;
    this.onChange(this.values);
  };

  return Render;
}();

exports.default = function (output, valueMap) {
  return new Render(output, valueMap);
};
