// popmotion/lib/value/composite-value.js
'use strict';

exports.__esModule = true;

var _ = require('./');

var _2 = _interopRequireDefault(_);

var _utils = require('../inc/utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var getChildCurrent = function (child) {
  return child.get();
};
var getChildVelocity = function (child) {
  return child.getVelocity();
};
var stopChild = function (child) {
  return child.stop();
};

var CompositeValue = function (_Value) {
  _inherits(CompositeValue, _Value);

  function CompositeValue(values, types) {
    _classCallCheck(this, CompositeValue);

    var _this = _possibleConstructorReturn(this, _Value.call(this));

    _this.types = types;
    _this.current = {};
    _this.velocity = {};
    _this.children = [];
    _this.update(values);
    return _this;
  }

  CompositeValue.prototype.update = function update(values) {
    for (var key in values) {
      var value = values[key];

      if (!this.hasOwnProperty(key)) {
        this.createChild(key, value);
      } else {
        this[key].update(values[key]);
      }

      this.current[key] = this[key].get();
    }

    this.__fireListeners();
  };

  CompositeValue.prototype.createChild = function createChild(key, value) {
    if ((0, _utils.isNum)(value)) {
      var type = this.types[key];
      this[key] = (0, _2.default)(value, type);
    } else if ((0, _utils.isString)(value)) {
      // Auto detect type?
    } else {
      this[key] = value;
    }

    this.children.push(key);
  };

  CompositeValue.prototype.eachChild = function eachChild(func) {
    var acc = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];

    var numChildren = this.children.length;
    for (var i = 0; i < numChildren; i++) {
      var childKey = this.children[i];
      acc[childKey] = func(this[childKey]);
    }
    return acc;
  };

  CompositeValue.prototype.get = function get() {
    return this.eachChild(getChildCurrent, this.current);
  };

  CompositeValue.prototype.getVelocity = function getVelocity() {
    return this.eachChild(getChildVelocity, this.velocity);
  };

  CompositeValue.prototype.stop = function stop() {
    this.eachChild(stopChild);
    return this;
  };

  return CompositeValue;
}(_.Value);

exports.default = function (values, types) {
  return new CompositeValue(values, types);
};
