// popmotion/lib/value/index.js
"use strict";

exports.__esModule = true;

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Value = exports.Value = function () {
  function Value(initialValue, type) {
    _classCallCheck(this, Value);

    this.current = initialValue;
    this.type = type;
    this.action = null;
  }

  Value.prototype.update = function update(newValue) {
    this.current = newValue;
    this.__fireListeners();
  };

  Value.prototype.get = function get() {
    return this.current;
  };

  Value.prototype.getVelocity = function getVelocity() {
    return this.action ? this.action.getVelocity : 0;
  };

  Value.prototype.stop = function stop() {
    if (this.action) {
      this.action.stop();
      this.action = null;
    }

    return this;
  };

  Value.prototype.registerAction = function registerAction(action) {
    var _this = this;

    if (this.action) {
      this.action.stop();
    }

    this.action = action;

    // Potential todo: Check if source action is this.action
    return function (v) {
      return _this.update(v);
    };
  };

  Value.prototype.addListener = function addListener(callback) {
    if (!this.listeners) {
      this.listeners = [];
    }

    var listenerIndex = this.listeners.indexOf(callback);
    if (listenerIndex === -1) {
      this.listeners.push(callback);
    }

    return this;
  };

  Value.prototype.removeListener = function removeListener(callback) {
    if (this.listeners) {
      var listenerIndex = this.listeners.indexOf(callback);
      if (listenerIndex > -1) {
        this.listeners.splice(listenerIndex, 1);
      }
    }

    return this;
  };

  Value.prototype.__fireListeners = function __fireListeners() {
    if (this.listeners) {
      var numListeners = this.listeners.length;
      for (var i = 0; i < numListeners; i++) {
        this.listeners[i](this.current, this);
      }
    }
  };

  return Value;
}();

exports.default = function (initialValue) {
  return new Value(initialValue);
};
