// popmotion/lib/value/transformers.js
'use strict';

exports.__esModule = true;
exports.steppedMotion = exports.restrictBetween = exports.restrictOver = exports.restrictUnder = exports.interpolate = exports.compose = exports.appendUnit = undefined;

var _calc = require('../inc/calc');

/**
 * Append Unit
 * Creates a function that will append the unit to a given value
 * appendUnit('px')(20) -> '20px'
 * @param  {string} unit)
 * @return {number}
 */
var appendUnit = exports.appendUnit = function (unit) {
  return function (v) {
    return '' + v + unit;
  };
};

/**
 * Compose
 * Compose other transformers to run linearily
 * compose(min(20), max(40))
 * @param  {...functions} transformers
 * @return {function}
 */
var compose = exports.compose = function () {
  for (var _len = arguments.length, transformers = Array(_len), _key = 0; _key < _len; _key++) {
    transformers[_key] = arguments[_key];
  }

  var numTransformers = transformers.length;
  var i = 0;

  return function (v, value) {
    for (i = 0; i < numTransformers; i++) {
      v = transformers[i](v, value);
    }

    return v;
  };
};

var interpolate = exports.interpolate = function (input, output, rangeEasing) {
  var rangeLength = input.length;
  var finalIndex = rangeLength - 1;

  return function (v) {
    // If value outside minimum range, quickly return
    if (v <= input[0]) {
      return output[0];
    }

    // If value outside maximum range, quickly return
    if (v >= input[finalIndex]) {
      return output[finalIndex];
    }

    var i = 0;

    // Find index of range start
    for (; i < rangeLength; i++) {
      if (input[i] > v || i === finalIndex) {
        break;
      }
    }

    var progressInRange = (0, _calc.getProgressFromValue)(v, input[i], input[i + 1]);
    var easedProgress = rangeEasing ? rangeEasing[i](progressInRange) : progressInRange;
    return (0, _calc.getValueFromProgress)(easedProgress, output[i], output[i + 1]);
  };
};

/**
 * Restrict value between
 * Creates a function that will restrict a given value between `min` and `max`
 * @param  {number} min
 * @param  {number} max
 * @return {number}
 */
var restrictUnder = exports.restrictUnder = function (max) {
  return function (v) {
    return Math.min(v, max);
  };
};
var restrictOver = exports.restrictOver = function (min) {
  return function (v) {
    return Math.max(v, min);
  };
};
var restrictBetween = exports.restrictBetween = function (min, max) {
  return compose(restrictOver(min), restrictUnder(max));
};

var steppedMotion = exports.steppedMotion = function (min, max) {
  return function (v) {};
};
