// popmotion/lib/value/types.js
'use strict';

exports.__esModule = true;
exports.complexString = exports.color = exports.hsla = exports.hex = exports.rgba = exports.rgbUnit = exports.scale = exports.number = exports.px = exports.percent = exports.degrees = exports.alpha = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _childTypes, _childTypes2; /**
                                * Value types
                                *
                                * alpha
                                * degrees
                                * hex
                                * hsla
                                * percent
                                * px
                                * rgbUnit
                                * rgb
                                * scale
                                */


var _filters = require('./filters');

var _utils = require('./utils');

var _utils2 = require('../inc/utils');

// String properties
var RED = 'Red';
var GREEN = 'Green';
var BLUE = 'Blue';
var ALPHA = 'Alpha';
var HUE = 'Hue';
var SATURATION = 'Saturation';
var LIGHTNESS = 'Lightness';

// Templates
var rgbaTemplate = function (colors) {
  return 'rgba(' + colors[RED] + ', ' + colors[GREEN] + ', ' + colors[BLUE] + ', ' + colors[ALPHA] + ')';
};
var hslaTemplate = function (colors) {
  return 'hsla(' + colors[HUE] + ', ' + colors[SATURATION] + ', ' + colors[LIGHTNESS] + ', ' + colors[ALPHA] + ')';
};

// Value types
var alpha = exports.alpha = {
  filter: (0, _filters.restrictBetween)(0, 1)
};

var degrees = exports.degrees = (0, _utils.createUnitType)('deg');
var percent = exports.percent = (0, _utils.createUnitType)('%');
var px = exports.px = (0, _utils.createUnitType)('px');

var number = exports.number = { test: _utils2.isNum };

var scale = exports.scale = {
  default: 1
};

var rgbUnit = exports.rgbUnit = {
  filter: (0, _filters.compose)(
  //http://codepen.io/osublake/full/xGVVaN/
  // (v, value, action) => {
  //   if (action) {
  //     const fromColor = action.from * action.from;
  //     const toColor = action.to * action.to;
  //     return Math.sqrt(action.progress * (toColor - fromColor) + fromColor);
  //   }

  //   return v;
  // },
  (0, _filters.restrictBetween)(0, 255), Math.round)
};

var rgba = exports.rgba = {
  childTypes: (_childTypes = {}, _childTypes[RED] = rgbUnit, _childTypes[GREEN] = rgbUnit, _childTypes[BLUE] = rgbUnit, _childTypes[ALPHA] = alpha, _childTypes),

  test: (0, _utils.isFirstChars)('rgb'),

  parse: (0, _utils.splitColorValues)([RED, GREEN, BLUE, ALPHA]),

  combine: rgbaTemplate
};

var hex = exports.hex = _extends({}, rgba, {

  test: (0, _utils.isFirstChars)('#'),

  parse: function (v) {
    var r = void 0,
        g = void 0,
        b = void 0;

    // If we have 6 characters, ie #FF0000
    if (v.length > 4) {
      r = v.substr(1, 2);
      g = v.substr(3, 2);
      b = v.substr(5, 2);

      // Or we have 3 characters, ie #F00
    } else {
      r = v.substr(1, 1);
      g = v.substr(2, 1);
      b = v.substr(3, 1);
      r += r;
      g += g;
      b += b;
    }

    return {
      Red: parseInt(r, 16),
      Green: parseInt(g, 16),
      Blue: parseInt(b, 16),
      Alpha: 1
    };
  }
});

var hsla = exports.hsla = {
  childTypes: (_childTypes2 = {}, _childTypes2[HUE] = number, _childTypes2[SATURATION] = percent, _childTypes2[LIGHTNESS] = percent, _childTypes2[ALPHA] = alpha, _childTypes2),

  test: (0, _utils.isFirstChars)('hsl'),

  parse: (0, _utils.splitColorValues)([HUE, SATURATION, LIGHTNESS, ALPHA]),

  combine: hslaTemplate
};

var color = exports.color = {
  childTypes: _extends({}, hsla.childTypes, rgba.childTypes),

  test: function (value) {
    return rgba.test(value) || hex.test(value) || hsla.test(value);
  }
};

var FLOAT_REGEX = /(-)?(\d[\d\.]*)/g;
var generateToken = function (token) {
  return '${' + token + '}';
};

var complexString = exports.complexString = {};
