// popmotion/lib/value/utils.js
'use strict';

exports.__esModule = true;
exports.createDelimited = exports.getColorValues = exports.findValueAndUnit = exports.createUnitType = exports.isFirstChars = exports.contains = undefined;
exports.splitColorValues = splitColorValues;

var _utils = require('../inc/utils');

var _filters = require('./filters');

/**
 *  Returns a function that will check any argument for `term`
 * `contains('needle')('haystack')`
 */
var contains = exports.contains = function (term) {
  return function (v) {
    return (0, _utils.isString)(term) && v.indexOf(term) !== -1;
  };
};

/**
 *  Returns a function that will check to see if an argument is
 *  the first characters in the provided `term`
 * `isFirstChars('needle')('haystack')`
 */
var isFirstChars = exports.isFirstChars = function (term) {
  return function (v) {
    return (0, _utils.isString)(term) && v.indexOf(term) === 0;
  };
};

/**
 * Create a unit value type
 */
var createUnitType = exports.createUnitType = function (type) {
  return {
    test: contains(type),
    parse: parseFloat,
    filter: (0, _filters.appendUnit)(type)
  };
};

/**
 * Creates a function that will split color
 * values from string into an object of properties
 * `splitColorValues(['Red', 'Green', 'Blue'])('rgba(0,0,0)')`
 */
function splitColorValues(terms) {
  var numTerms = terms.length;

  return function (v) {
    var values = {};
    var valuesArray = (0, _utils.splitCommaDelimited)((0, _utils.getValueFromFunctionString)(v));

    for (var i = 0; i < numTerms; i++) {
      values[terms[i]] = valuesArray[i] !== undefined ? parseFloat(valuesArray[i]) : 1;
    }

    return values;
  };
}

/*
  Split a value into a value/unit object
  
    "200px" -> { value: 200, unit: "px" }
    
  @param [string]: Value to split
  @return [object]: Object with value and unit props
*/
var findValueAndUnit = exports.findValueAndUnit = function (value) {
  if (value.match) {
    var splitValue = value.match(/(-?\d*\.?\d*)(.*)/);

    return {
      value: parseFloat(splitValue[1]),
      unit: splitValue[2]
    };
  } else {
    return { value: value };
  }
};

/*
  Split color string into map of color properties

  "rgba(255, 255, 255, 0)", ["Red", 'Green", "Blue", "Alpha"]

  { Red: 255... }
*/
var getColorValues = exports.getColorValues = function (value, colorTerms) {
  var numColorTerms = colorTerms.length;
  var colorValues = {};
  var colors = (0, _utils.splitCommaDelimited)((0, _utils.getValueFromFunctionString)(value));

  for (var i = 0; i < numColorTerms; i++) {
    colorValues[colorTerms[i]] = colors[i] !== undefined ? colors[i] : 1;
  }

  return colorValues;
};

/*
  Get value from function string

  "translateX(20px)" -> "20px"
*/
// export const getValueFromFunctionString = (value) => value.substring(value.indexOf('(') + 1, value.lastIndexOf(')'));


var createDelimited = exports.createDelimited = function (values, terms, delimiter, chop) {
  var numTerms = terms.length;
  var combined = '';

  for (var i = 0; i < numTerms; i++) {
    var term = terms[i];
    if (values.hasOwnProperty(term)) {
      combined += values[term] + delimiter;
    }
  }

  if (chop) {
    combined = combined.slice(0, -chop);
  }

  return combined;
};
