// popmotion/lib/_input/Gyro.js
'use strict';

exports.__esModule = true;

var _Pointer = require('./Pointer');

var _Pointer2 = _interopRequireDefault(_Pointer);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var DEVICE_MOTION = 'devicemotion';

/*
  Scrape x/y coordinates from provided event

  @param [event]
  @return [object]
*/
var eventToPoint = function (e) {
  return {
    x: e.accelerationIncludingGravity.x,
    y: e.accelerationIncludingGravity.y,
    z: e.accelerationIncludingGravity.z,
    yaw: e.alpha,
    pitch: e.beta,
    roll: event.gamma
  };
};

var Gyro = function (_Input) {
  _inherits(Gyro, _Input);

  function Gyro() {
    _classCallCheck(this, Gyro);

    var _this = _possibleConstructorReturn(this, _Input.call(this));

    _this.bindEvents();
    return _this;
  }

  Gyro.prototype.latest = function latest(e) {
    _Input.prototype.latest.call(this, eventToPoint(e));
    e.preventDefault();
  };

  Gyro.prototype.start = function start() {
    _Input.prototype.start.call(this);
    document.documentElement.addEventListener(DEVICE_MOTION, this.onMove);
  };

  Gyro.prototype.stop = function stop() {
    _Input.prototype.stop.call(this);
    document.documentElement.removeEventListener(DEVICE_MOTION, this.onMove);
  };

  return Gyro;
}(_Pointer2.default);

exports.default = Gyro;
